include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,814 @@
+  // findOrConstruct returns a reference to the value stored in any of the
+  // internal map containers or constructs a new value.
+  // -------------------------------------------------------------
+
+  template <typename OpT> ValueT &findOrConstruct() {
+    return findOrConstruct(OpDescription::get<OpT>());
+  }
+
+  // Return a reference to the operation described by a given OpDescription.
+  // This includes any of the possible operation kinds.
+  // Constructs a new entry if the operation was found in neither map.
+  ValueT &findOrConstruct(const OpDescription &desc) {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      assert(desc.getOpcodes().size() == 1 &&
+             "OpMap only supports querying of single core opcodes and "
+             "intrinsics.");
+
+      const unsigned op = desc.getOpcode();
+
+      if (desc.isCoreOp())
+        return m_coreOpcodes[op];
+
+      if (desc.isIntrinsic())
+        return m_intrinsics[op];
+    }
+
+    if (desc.isDialectOp()) {
+      for (DialectOpKVT &dialectOpKV : m_dialectOps) {
+        if (dialectOpKV == desc)
+          return dialectOpKV.Value;
+      }
+    }
+
+    // Nothing found. Insert a new dialect operation.
+    return insertDialectOp(desc, {});
+  }
+
+  // -------------------------------------------------------------
+  // Convenience operator[] definitions.
+  // -------------------------------------------------------------
+
+  ValueT &operator[](const OpDescription &desc) {
+    return findOrConstruct(desc);
+  }
+  ValueT *operator[](const Function &func) { return findPointer(func); }
+  ValueT *operator[](const Instruction &inst) { return findPointer(inst); }
+
+  // -------------------------------------------------------------
+  // lookup tries to find whether a given function or instruction
+  // can be mapped to any of the entries in the internal map
+  // containers.
+  // It returns either a default-constructed object if the key
+  // was not found or a copy of the contained value.
+  // -------------------------------------------------------------
+
+  // Try to lookup a function which is either the callee of an intrinsic call
+  // or a dialect operation.
+  ValueT lookup(const Function &func) const {
+    auto it = find(func);
+    if (auto val = it.val(); val)
+      return *val;
+
+    return {};
+  }
+
+  // Try to lookup an instruction which is either an intrinsic instruction,
+  // a dialect operation or a core instruction.
+  ValueT lookup(const Instruction &inst) const {
+    auto it = find(inst);
+    if (auto val = it.val(); val)
+      return *val;
+
+    return {};
+  }
+
+  // -------------------------------------------------------------
+  // Inserts a given Key/Value pair and return a reference or
+  // directly return a reference if the key already exists.
+  // -------------------------------------------------------------
+
+  ValueT &findOrConstruct(const OpDescription &desc, const ValueT &val) {
```

Comment:
This API is too surprising. When you say "find or construct" I hear "either find (and leave the map unmodified) or construct a new element". What this method does instead is to unconditionally assign `val` to `desc`.

It seems this is primarily required for the copy constructor. Could you instead just implement `try_emplace`? That is:
```c++
  template <typename... Ts>
  std::pair<iterator, bool> try_emplace(const OpDescription &desc, Ts &&... args) {
    ...
    auto [it, inserted] = m_coreOpcodes.try_emplace(op, std::forward<Ts>(args)...);
    return {iterator_t(this, it, false), inserted};
    ...
  }
```

Reply:
Yeah, makes sense. This requires some adjustments. Generally, an issue for the Visitor refactoring is that we constrained the `OpDescriptions` to be inserted to be single-opcodes. This means, something like that:

```
      m_opMap.insert(try_emplace, {handlerIdx});
```

does not neccessarily work.

I'd rather remove that constraint for insertion and return an iterator range instead. Same goes for the `OpMapIterator::erase()` method, since it is rather inefficient to only erase a single operation at a time.

Reply:
Hmm, but if the OpDescription has multiple opcodes, then inserted locations don't necessarily fall into a contiguous range of iterators.

For the longer term, I'd rather remove the support for multi-opcodes/multi-intrinsics in OpDescription. It's always been more of a hack.